### PR TITLE
feat: display customer reviews on product detail page

### PR DESCRIPTION
**Pull request by paramaeder**
## PR Type

[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no API changes)
[ ] Build-related changes
[ ] CI-related changes
[ ] Documentation content changes
[ ] Application / infrastructure changes
[ ] Other: 

## What Is the Current Behavior?

Currently there is no feature to display customer reviews

Issue Number: Closes 76116

## What Is the New Behavior?

If there is at least one product review then there is a reviews tab shown in the product-detail-info.component which can show customer reviews

## Does this PR Introduce a Breaking Change?

[ ] Yes
[x] No

### CODE REVIEW

**Comment by SGrueber** [comment]
The ui looks fine, just 2 hints: The heading for the review display should be 'Recent Product Reviews' . If there are no reviews to display the message 'There are no reviews submitted yet' should be shown. Please, check also the provided IAD. https://www.figma.com/file/6Ccec915dDuc2DIMf89ALj/Product-Reviews?node-id=1%3A448

File: src/app/core/services/reviews/reviews.service.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  getProductReviews(sku: string): Observable<ProductReviews> {
+    if (!sku) {
+      return throwError(() => new Error('getProductReviews() called without sku'));
+    }
+
+    let params = new HttpParams();
+    params.append('sku', sku);
+
+    return this.currentCustomer$.pipe(
+      tap(customer => {
```
Comment (SGrueber) [inline src/app/core/services/reviews/reviews.service.ts:29]: Does the review REST api really require this? If not, remove the customer selection stuff

File: src/app/core/models/product-reviews/product-reviews.mapper.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import { Injectable } from '@angular/core';
+
+import { ProductReviewData } from './product-reviews.interface';
+import { ProductReviews } from './product-reviews.model';
+
+@Injectable({ providedIn: 'root' })
+export class ProductReviewsMapper {
+  static fromData(productReviewsData: ProductReviewData[], sku: string): ProductReviews {
```
Comment (SGrueber) [inline src/app/core/models/product-reviews/product-reviews.mapper.ts:8]: I'd prefer the order sku, productReviews

File: src/app/core/services/reviews/reviews.service.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+  private currentCustomer$ = this.store.pipe(select(getLoggedInCustomer), take(1));
+
+  getProductReviews(sku: string): Observable<ProductReviews> {
+    if (!sku) {
+      return throwError(() => new Error('getProductReviews() called without sku'));
+    }
+
+    let params = new HttpParams();
+    params.append('sku', sku);
```
Comment (SGrueber) [inline src/app/core/services/reviews/reviews.service.ts:26]: is this necessary? the sku ist part of the url?

File: src/app/core/store/shopping/products/products.effects.ts [hunk 4]
```diff
@@ -109,6 +111,26 @@ export class ProductsEffects {
...
+    this.actions$.pipe(
+      ofType(loadProductSuccess),
+      mapToPayloadProperty('product'),
+      mapToProperty('sku'),
+      whenTruthy(),
+      map(sku => loadProductPrices({ skus: [sku] }))
+    )
+  );
+
+  loadProductReviewsAfterProductSuccess$ = createEffect(() =>
```
Comment (SGrueber) [inline src/app/core/store/shopping/products/products.effects.ts:124]: loading of the product review should be triggered by the product-reviews. component with the help of a facade method

File: src/app/core/store/shopping/product-reviews/product-reviews.actions.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import { createAction } from '@ngrx/store';
+
+import { ProductReviews } from 'ish-core/models/product-reviews/product-reviews.model';
+import { payload } from 'ish-core/utils/ngrx-creators';
+
+export const loadProductReviews = createAction('[Product Review] Load Product Reviews', payload<{ skus: string[] }>());
```
Comment (SGrueber) [inline src/app/core/store/shopping/product-reviews/product-reviews.actions.ts:6]: we don't need to fetch the reviews of more than one product so we the parameter should be sku of type string

**Review by SGrueber — CHANGES_REQUESTED** [review]

File: src/app/shared/components/product/product-reviews/product-reviews.component.html [hunk 6]
```diff
@@ -0,0 +1,24 @@
+<h3>{{ 'product.reviews.section.heading' | translate | uppercase }}</h3>
+
+<div *ngIf="(productReviews$ | async)?.length > 0; else noReviews">
+  <div *ngFor="let review of productReviews$ | async">
+    <div class="row pb-3">
```
Comment (SGrueber) [inline src/app/shared/components/product/product-reviews/product-reviews.component.html:5]: add class 'review-item' here

File: src/app/shared/components/product/product-reviews/product-reviews.component.html [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+      <div class="col-md-3 text-muted small">
+        {{ review.authorFirstName }}
+        <br />
+        {{ review.creationDate | date: 'MMM d, y, h:mm:ss a' }}
+      </div>
+      <div class="col-md-9">
+        <ng-container *ngFor="let fill of stars$ | async; let lastStar = last">
+          <ish-product-rating-star [filled]="fill" [lastStar]="lastStar"></ish-product-rating-star>
+          <span *ngIf="lastStar" class="mb-0 h5 font-weight-light"> {{ review.title }}</span>
```
Comment (SGrueber) [inline src/app/shared/components/product/product-reviews/product-reviews.component.html:14]: use class 'review-item-header-title' here - define further styles in the ratings.scss if needed

File: src/app/shared/components/product/product-reviews/product-reviews.component.spec.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    fixture = TestBed.createComponent(ProductReviewsComponent);
+    component = fixture.componentInstance;
+    element = fixture.nativeElement;
+  });
+
+  it('should be created', () => {
+    expect(component).toBeTruthy();
+    expect(element).toBeTruthy();
+    expect(() => fixture.detectChanges()).not.toThrow();
+  });
```
Comment (SGrueber) [inline src/app/shared/components/product/product-reviews/product-reviews.component.spec.ts:37]: maybe add another test here to check that a review is displayed

**Review by SGrueber — CHANGES_REQUESTED** [review]

File: src/app/core/facades/shopping.facade.ts [hunk 9]
```diff
@@ -222,4 +223,10 @@ export class ShoppingFacade {
     });
     return this.store.pipe(select(getPromotions(promotionIds)));
   }
+
+  // REVIEWS
+  getProductReviews$(sku: string) {
+    this.store.dispatch(loadProductReviews({ sku }));
```
Comment (SGrueber) [inline src/app/core/facades/shopping.facade.ts:229]: I suggest to check here if the feature is enabled - so an action LoadProductReviews is never called if the feature is off:
 getProductReviews$(sku: string) {
    if (this.featureToggleService.enabled('rating')) {
      this.store.dispatch(loadProductReviews({ sku }));
    }
    return this.store.pipe(select(getProductReviewsBySku(sku)));
  }

**Review by SGrueber — CHANGES_REQUESTED** [review]

File: src/app/shared/components/product/product-reviews/product-reviews.component.html [hunk 10]
```diff
@@ -0,0 +1,24 @@
+<h3>{{ 'product.reviews.section.heading' | translate | uppercase }}</h3>
+
+<div *ngIf="(productReviews$ | async)?.length > 0; else noReviews">
+  <div *ngFor="let review of productReviews$ | async">
+    <div class="row pb-3">
+      <div class="col-md-3 text-muted small">
+        {{ review.authorFirstName }}
+        <br />
+        {{ review.creationDate | date: 'MMM d, y, h:mm:ss a' }}
```
[the remark this answers is not in the document]
Comment (paramaeder) [reply src/app/shared/components/product/product-reviews/product-reviews.component.html:9]: Is it possible that the pipe can only display the date, but not the time?

[the remark this answers is not in the document]
Comment (SGrueber) [reply src/app/shared/components/product/product-reviews/product-reviews.component.html:9]: You can use all format options of the angular date pipe: https://angular.io/api/common/DatePipe
This should be the one you need:  {{ review.creationDate | ishDate: 'medium' }}

File: src/app/core/services/reviews/reviews.service.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+   *
+   * @param sku The product sku
+   * @returns   The available reviews of a product
+   */
+  getProductReviews(sku: string): Observable<ProductReviews> {
+    if (!sku) {
+      return throwError(() => new Error('getProductReviews() called without sku'));
+    }
+
+    return this.currentCustomer$.pipe(
```
Comment (SGrueber) [inline src/app/core/services/reviews/reviews.service.ts:30]: this is still not necessary: return this.apiService... should be enough

File: src/app/core/store/shopping/product-reviews/product-reviews.reducer.ts [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+const initialState: ProductReviewsState = productReviewsAdapter.getInitialState({
+  loading: false,
+  error: undefined,
+});
+
+export const productReviewsReducer = createReducer(
+  initialState,
+  on(loadProductReviews, state => ({ ...state, loading: true })),
```
Comment (SGrueber) [inline src/app/core/store/shopping/product-reviews/product-reviews.reducer.ts:26]: pls use the methods setLoadingOn, unsetLoadingAndErrorOn, 

make sure, loading is set by the load action, and unset by the fail and success action and the error is reseted at least by the next success action

**Review by SGrueber — CHANGES_REQUESTED** [review]

File: src/app/core/store/shopping/product-reviews/product-reviews.effects.spec.ts [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+      actions$ = of(action);
+
+      effects.loadProductReviews$.subscribe(() => {
+        const sku = capture(reviewsServiceMock.getProductReviews);
+        expect(sku).toEqual(sku);
+        done();
+      });
+    });
+
+    it('should map invalid request to action of type loadProductReviewsFail', () => {
```
Comment (SGrueber) [inline src/app/core/store/shopping/product-reviews/product-reviews.effects.spec.ts:46]: Usually we also test the the mapping to the success action

File: src/app/shared/components/product/product-reviews/product-reviews.component.ts [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+  constructor(private context: ProductContextFacade, private shoppingFacade: ShoppingFacade) {}
+
+  ngOnInit() {
+    this.productReviews$ = this.context.select('reviews');
+    this.rating$ = this.context.select('product', 'roundedAverageRating').pipe(filter(x => typeof x === 'number'));
+    this.stars$ = this.rating$.pipe(
+      map(rate => range(1, 6).map(index => (index <= rate ? 'full' : index - 0.5 === rate ? 'half' : 'empty')))
+    );
+    this.error$ = this.context.productReviewsError$;
```
Comment (SGrueber) [inline src/app/shared/components/product/product-reviews/product-reviews.component.ts:31]: use the method of the shopping facade shopping facade here

**Review by dhhyi — CHANGES_REQUESTED** [review]
This feature should be externalized into an extension. I doubt every shop will use it.

File: src/assets/i18n/fr_FR.json [hunk 15]
```diff
@@ -891,6 +891,9 @@
   "product.ready_for_shipment.within24": "Habituellement prêt à être expédié dans les 24 heures",
   "product.remove_all_product_filters.text": "Réinitialiser tous les filtres",
   "product.retailSet.products.heading": "Ce jeu de produits contient",
+  "product.reviews.heading": "Notations",
```
Comment (mglatter) [inline src/assets/i18n/fr_FR.json:894]: _Reviews_ and _submit_ are translated differently according to our translation memory system. Please ask Nicolas for review of fr_FR.json

**Review by mglatter — COMMENTED** [review]
Approved for DE and EN JSON files, for FR please ask Nicolas for a review.

**Comment by SGrueber** [comment]
the changes are part of the pr #1174

**Pull request closed without merging**